### PR TITLE
implement tests for modal rendering

### PR DESCRIPTION
**Pull request by surekhaw**
## Description
tests for modal rendering - render value for current product, for related product, for when features are unique and in common, as well as rendering check marks for when value is boolean true.

## How to test
run npm test

## Notes
starting to appreciate the concept of testing for more than just passing assignments. Writing these tests actually helped me find a couple of bugs, which are now fixed.

## Issue tracking
https://trello.com/c/OVHhJzcv/127-m-write-tests-for-modal-rendering

### CODE REVIEW

File: client/src/components/relatedProducts/Modal.jsx [hunk 1]
```diff
@@ -9,38 +10,69 @@ const Modal = (props) => {
...
-      <tr className='features'>
-        <td className='check'>{features[key].currentProduct}</td>
-        <td>{key}</td>
-        <td className='check'>{features[key].comparedProduct}</td>
+  const rows = keys.map((key) => {
+    if (features[key].currentProduct === true && features[key].comparedProduct === true) {
+      return (
+        <tr className='features'>
+          <td className='check'>
+            <FontAwesomeIcon data-testid='checkMark' icon={ faCheck } />
```
Comment (tm2838) [inline client/src/components/relatedProducts/Modal.jsx:38]: If the icon is added to the library then here it can be referred to as `<FontAwesomeIcon data-testid='checkMark' icon='check' />` I believe? and we wouldn't need the import on line 2. (learned from @shabbyblue16  lol)

File: client/src/components/relatedProducts/Modal.jsx [hunk 2]
```diff
@@ -9,38 +10,69 @@ const Modal = (props) => {
...
   const keys = Object.keys(features);
   // eslint-disable-next-line react/jsx-key
-  const rows = keys.map((key) => (
-      // eslint-disable-next-line react/jsx-key
-      <tr className='features'>
-        <td className='check'>{features[key].currentProduct}</td>
-        <td>{key}</td>
-        <td className='check'>{features[key].comparedProduct}</td>
+  const rows = keys.map((key) => {
+    if (features[key].currentProduct === true && features[key].comparedProduct === true) {
```
Comment (tm2838) [inline client/src/components/relatedProducts/Modal.jsx:34]: Just my preference though:

```suggestion
    if (features[key].currentProduct && features[key].comparedProduct) {
```

File: client/src/components/relatedProducts/Modal.jsx [hunk 3]
```diff
@@ -9,38 +10,69 @@ const Modal = (props) => {
...
         currentProduct: currentProduct.features[i].value,
       };
     }
     // eslint-disable-next-line no-plusplus
     for (let j = 0; j < comparedProduct.product.features.length; j++) {
-      if (comparedProduct.product.features[j].value === true) {
-        comparedProduct.product.features[j].value = <FontAwesomeIcon icon={'faCheck'} />;
-      } else if (comparedProduct.product.features[j].value === false) {
-        comparedProduct.product.features[j].value = '';
+      if (!features[comparedProduct.product.features[j].feature]) {
```
Comment (tm2838) [inline client/src/components/relatedProducts/Modal.jsx:19]: assigning `comparedProduct.product.features[j].feature` to a variable and refer to that variable afterwards may help? feel like needing to write it out every time for me is a pain 🤔

File: client/src/components/relatedProducts/Modal.jsx [hunk 4]
```diff
@@ -9,38 +10,69 @@ const Modal = (props) => {
...
-        currentProduct.features[i].value = <FontAwesomeIcon icon={'faCheck'} />;
-      } else if (currentProduct.features[i].value === false) {
-        currentProduct.features[i].value = '';
-      }
       features[currentProduct.features[i].feature] = {
         currentProduct: currentProduct.features[i].value,
       };
     }
     // eslint-disable-next-line no-plusplus
     for (let j = 0; j < comparedProduct.product.features.length; j++) {
```
Comment (tm2838) [inline client/src/components/relatedProducts/Modal.jsx:18]: eslint doesn't like `i++` but won't complain about`i+=1`, weird, lol

File: client/src/components/relatedProducts/Modal.jsx [hunk 5]
```diff
@@ -9,38 +10,69 @@ const Modal = (props) => {
...
 
   const keys = Object.keys(features);
   // eslint-disable-next-line react/jsx-key
-  const rows = keys.map((key) => (
-      // eslint-disable-next-line react/jsx-key
-      <tr className='features'>
-        <td className='check'>{features[key].currentProduct}</td>
-        <td>{key}</td>
-        <td className='check'>{features[key].comparedProduct}</td>
+  const rows = keys.map((key) => {
```
Comment (tm2838) [inline client/src/components/relatedProducts/Modal.jsx:33]: It seems `rows` can be extracted as a separate component almost? since the return values are all elements

File: client/src/components/relatedProducts/test/modal.test.js [hunk 6]
```diff
@@ -0,0 +1,119 @@
+/* eslint-disable no-unused-vars */
+import React from 'react';
+import { render } from '@testing-library/react';
+import Modal from '../Modal.jsx';
+
+describe('modal', () => {
+  const currentProduct = {
```
Comment (tm2838) [inline client/src/components/relatedProducts/test/modal.test.js:7]: feel free to import from `testProduct` or add more tailored mock data, which may save a few lines for the test file

**Review by tm2838 — APPROVED** [review]
Thanks for adding more tests and sorry for leaving these comments lol! None of them is blocking though 👍

File: client/src/components/relatedProducts/Modal.jsx (hunk 1 above)
Comment (surekhaw) [reply client/src/components/relatedProducts/Modal.jsx:38]: I can't check manually until we have either a current or related product with the appropriate value.  The tests only check for the testId. Will flag this for potential future refactoring.

File: client/src/components/relatedProducts/Modal.jsx (hunk 2 above)
Comment (surekhaw) [reply client/src/components/relatedProducts/Modal.jsx:34]: This was intentional because I don't want a truthy. I only want the check to render if the value is a boolean, specifically true.

File: client/src/components/relatedProducts/Modal.jsx (hunk 3 above)
Comment (surekhaw) [reply client/src/components/relatedProducts/Modal.jsx:19]: if there's time, I'll flag for refactoring.

File: client/src/components/relatedProducts/Modal.jsx (hunk 5 above)
Comment (surekhaw) [reply client/src/components/relatedProducts/Modal.jsx:33]: great catch!  totally missed that in my refactoring from the skeleton to be dynamic and account for different characteristics.  Again, will flag for refactor based on time.

**Comment by surekhaw** [comment]
Love the comments!  Thank you so much. It's so helpful to get a fresh outlook.  I feel kind of under the gun after that lecture, so I will definitely come back and refactor if there's time. For now, it's working, so I want to make sure I get everything done (or as close as possible).
